allennlp/interpret/attack/hotflip.py

```diff
@@ -0,0 +1,184 @@
+                        outputs[each] = outputs[each].detach().cpu().numpy().squeeze().squeeze()
+                    elif isinstance(outputs[each],list):                    
+                        outputs[each] = outputs[each][0]                        
+                self.predictor.predictions_to_labeled_instances(new_instances[0], outputs)                
+
+                # if the prediction has changed, then stop
+                if "tags" in new_instances[0]: # NER case                
+                    cur_tag_dict = defaultdict(int)
+                    cur_tag_tok = ''
+                    for label in new_instances[0]["tags"].__dict__["field_list"]                    :
+                        if label.label != "O":
+                            cur_tag_dict[label.label] += 1
+                            cur_tag_tok = cur_tag_tok + label.label                                        
+                    if not (cur_tag_dict == tag_dict) and (cur_tag_tok == tag_tok):
+                        break
+                else:  
+                    label_change = False                  
+                    for field in check_list.keys():                        
+                        if field in new_instances[0].fields:
+                            equal = new_instances[0][field].__eq__(check_list[field])                            
+                        else:
+                            equal = outputs[field] == check_list[field]                            
+                        # Save prediction                    
+                        if not equal: 
+                            label_change = True
+                            break                                                   
+                    if label_change:                    
+                        if "probs" in outputs: # sentiment analysis
+                            new_prediction = outputs["probs"]                    
+                        elif "label_probs" in outputs: # textual entailment
+                            new_prediction = outputs["label_probs"]                    
+                        elif "best_span_str" in outputs: # bidaf
+                            new_prediction = outputs["best_span_str"]
+                        elif "answer" in outputs: # NAQANet
+                            ans_type = outputs["answer"]["answer_type"]
+                            if ans_type == "count":
+                                new_prediction = outputs["answer"]["count"]
+                            else:
+                                new_prediction = outputs["answer"]["value"]
+                        break
+                        
+            final_tokens.append(current_tokens)        
+        return sanitize({"final": final_tokens,"original": original, "new_prediction": new_prediction})
+        
+    def first_order_taylor(self, grad, embedding_matrix, token_idx):            
+        """
+        the below code is based on
```

Comment:
This is good to have, but I'd also include an English description of what the logic does.